File: sdk/typescript/src/namespaces/usage-metering.ts
/**
 * Usage Metering Namespace API
 *
 * Provides usage tracking and billing information:
 * - Usage summaries and breakdowns
 * - Usage limits and quotas
 * - Export usage data
 */

/**
 * Usage period options.
 */
export type UsagePeriod = 'hour' | 'day' | 'week' | 'month' | 'quarter' | 'year';

/**
 * Billing tier.
 */
export type BillingTier = 'free' | 'starter' | 'professional' | 'enterprise' | 'enterprise_plus';

/**
 * Export format.
 */
export type UsageExportFormat = 'csv' | 'json';

/**
 * Token usage.
 */
export interface TokenUsage {
  input: number;
  output: number;
  total: number;
  cost: string;
}

/**
 * Usage counts.
 */
export interface UsageCounts {
  debates: number;
  api_calls: number;
}

/**
 * Usage summary.
 */
export interface UsageSummary {
  period_start: string;
  period_end: string;
  period_type: UsagePeriod;
  tokens: TokenUsage;
  counts: UsageCounts;
  by_model: Record<string, number>;
  by_provider: Record<string, number>;
  limits: {
    tokens: number;
    debates: number;
    api_calls: number;
  };
  usage_percent: {
    tokens: number;
    debates: number;
    api_calls: number;
  };
}

/**
 * Model usage breakdown.
 */
export interface ModelUsage {
  model: string;
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  cost: string;
  requests: number;
}

/**
 * Provider usage breakdown.
 */
export interface ProviderUsage {
  provider: string;
  total_tokens: number;
  cost: string;
  requests: number;
}

/**
 * Daily usage breakdown.
 */
export interface DailyUsage {
  day: string;
  total_tokens: number;
  cost: string;
  debates: number;
  api_calls: number;
}

/**
 * User usage breakdown.
 */
export interface UserUsage {
  user_id: string;
  total_tokens: number;
  cost: string;
  requests: number;
}

/**
 * Full usage breakdown.
 */
export interface UsageBreakdown {
  org_id: string;
  period_start: string;
  period_end: string;
  totals: {
    cost: string;
    tokens: number;
    debates: number;
    api_calls: number;
  };
  by_model: ModelUsage[];
  by_provider: ProviderUsage[];
  by_day: DailyUsage[];
  by_user: UserUsage[];
}

/**
 * Usage limits.
 */
export interface UsageLimits {
  org_id: string;
  tier: BillingTier;
  limits: {
    tokens: number;
    debates: number;
    api_calls: number;
  };
  used: {
    tokens: number;
    debates: number;
    api_calls: number;
  };
  percent: {
    tokens: number;
    debates: number;
    api_calls: number;
  };
  exceeded: {
    tokens: boolean;
    debates: boolean;
    api_calls: boolean;
  };
}

/**
 * Quota period.
 */
export type QuotaPeriod = 'hour' | 'day' | 'week' | 'month';

/**
 * Quota status.
 */
export interface QuotaStatus {
  limit: number;
  current: number;
  remaining: number;
  period: QuotaPeriod;
  percentage_used: number;
  is_exceeded: boolean;
  is_warning: boolean;
  resets_at?: string;
}

/**
 * All quotas response.
 */
export interface QuotasResponse {
  quotas: {
    debates: QuotaStatus;
    api_requests: QuotaStatus;
    tokens: QuotaStatus;
    storage_bytes: QuotaStatus;
    knowledge_bytes: QuotaStatus;
  };
}

/**
 * Usage breakdown options.
 */
export interface UsageBreakdownOptions {
  start?: string;
  end?: string;
}

/**
 * Usage export options.
 */
export interface UsageExportOptions {
  start?: string;
  end?: string;
  format?: UsageExportFormat;
}

/**
 * Client interface for usage metering operations.
 */
interface UsageMeteringClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }
  ): Promise<T>;
}

/**
 * Usage Metering API namespace.
 *
 * Provides methods for tracking and managing usage:
 * - Get usage summaries and breakdowns
 * - Check usage limits and quotas
 * - Export usage data
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai' });
 *
 * // Get monthly usage summary
 * const { usage } = await client.usageMetering.getUsage('month');
 *
 * // Get detailed breakdown
 * const { breakdown } = await client.usageMetering.getBreakdown({
 *   start: '2025-01-01T00:00:00Z',
 *   end: '2025-01-31T23:59:59Z',
 * });
 *
 * // Check quotas
 * const { quotas } = await client.usageMetering.getQuotas();
 * ```
 */
export class UsageMeteringAPI {
  constructor(private client: UsageMeteringClientInterface) {}

  /**
   * Get usage summary for a billing period.
   */
  async getUsage(period: UsagePeriod = 'month'): Promise<{ usage: UsageSummary }> {
    return this.client.request('GET', '/api/v1/billing/usage', {
      params: { period },
    });
  }

  /**
   * Get detailed usage breakdown.
   */
  async getBreakdown(options?: UsageBreakdownOptions): Promise<{ breakdown: UsageBreakdown }> {
    return this.client.request('GET', '/api/v1/billing/usage/breakdown', {
      params: options as Record<string, unknown>,
    });
  }

  /**
   * Get current usage limits.
   */
  async getLimits(): Promise<{ limits: UsageLimits }> {
    return this.client.request('GET', '/api/v1/billing/limits');
  }

  /**
   * Get quota status for all resources.
   */
  async getQuotas(): Promise<QuotasResponse> {
    return this.client.request('GET', '/api/v1/quotas');
  }

  /**
   * Export usage data as CSV or JSON.
   */
  async exportUsage(options?: UsageExportOptions): Promise<UsageBreakdown | string> {
    return this.client.request('GET', '/api/v1/billing/usage/export', {
      params: options as Record<string, unknown>,
    });
  }
}

File: sdk/python/aragora_sdk/namespaces/usage_metering.py
"""
Usage Metering Namespace API

Provides usage tracking and billing information:
- Usage summaries and breakdowns
- Usage limits and quotas
- Export usage data

Features:
- Track token and API usage
- Monitor billing limits
- Export reports
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


UsagePeriod = Literal["hour", "day", "week", "month", "quarter", "year"]
BillingTier = Literal["free", "starter", "professional", "enterprise", "enterprise_plus"]
UsageExportFormat = Literal["csv", "json"]


class UsageMeteringAPI:
    """
    Synchronous Usage Metering API.

    Provides methods for tracking and managing usage:
    - Get usage summaries and breakdowns
    - Check usage limits and quotas
    - Export usage data

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai", api_key="...")
        >>> usage = client.usage_metering.get_usage("month")
        >>> quotas = client.usage_metering.get_quotas()
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    def get_usage(self, period: UsagePeriod = "month") -> dict[str, Any]:
        """
        Get usage summary for a billing period.

        Args:
            period: Time period (hour, day, week, month, quarter, year)

        Returns:
            Dict with usage summary including:
            - period_start/period_end
            - tokens (input/output/total/cost)
            - counts (debates/api_calls)
            - by_model/by_provider breakdowns
            - limits and usage_percent
        """
        return self._client.request("GET", "/api/v1/billing/usage", params={"period": period})

    def get_breakdown(
        self,
        start: str | None = None,
        end: str | None = None,
    ) -> dict[str, Any]:
        """
        Get detailed usage breakdown.

        Args:
            start: Start date (ISO 8601)
            end: End date (ISO 8601)

        Returns:
            Dict with breakdown including:
            - totals (cost/tokens/debates/api_calls)
            - by_model: Model-level breakdown
            - by_provider: Provider-level breakdown
            - by_day: Daily breakdown
            - by_user: User-level breakdown
        """
        params: dict[str, Any] = {}
        if start:
            params["start"] = start
        if end:
            params["end"] = end
        return self._client.request(
            "GET", "/api/v1/billing/usage/breakdown", params=params if params else None
        )

    def get_limits(self) -> dict[str, Any]:
        """
        Get current usage limits.

        Returns:
            Dict with:
            - org_id: Organization ID
            - tier: Billing tier
            - limits: Token/debate/API call limits
            - used: Current usage
            - percent: Usage percentage
            - exceeded: Whether limits exceeded
        """
        return self._client.request("GET", "/api/v1/billing/limits")

    def get_quotas(self) -> dict[str, Any]:
        """
        Get quota status for all resources.

        Returns:
            Dict with quotas for:
            - debates
            - api_requests
            - tokens
            - storage_bytes
            - knowledge_bytes
        """
        return self._client.request("GET", "/api/v1/quotas")

    def export_usage(
        self,
        start: str | None = None,
        end: str | None = None,
        format: UsageExportFormat = "json",
    ) -> dict[str, Any] | str:
        """
        Export usage data as CSV or JSON.

        Args:
            start: Start date (ISO 8601)
            end: End date (ISO 8601)
            format: Export format (csv or json)

        Returns:
            Usage data in requested format
        """
        params: dict[str, Any] = {"format": format}
        if start:
            params["start"] = start
        if end:
            params["end"] = end
        return self._client.request("GET", "/api/v1/billing/usage/export", params=params)


class AsyncUsageMeteringAPI:
    """
    Asynchronous Usage Metering API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     usage = await client.usage_metering.get_usage("month")
        ...     breakdown = await client.usage_metering.get_breakdown()
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def get_usage(self, period: UsagePeriod = "month") -> dict[str, Any]:
        """Get usage summary for a billing period."""
        return await self._client.request("GET", "/api/v1/billing/usage", params={"period": period})

    async def get_breakdown(
        self,
        start: str | None = None,
        end: str | None = None,
    ) -> dict[str, Any]:
        """Get detailed usage breakdown."""
        params: dict[str, Any] = {}
        if start:
            params["start"] = start
        if end:
            params["end"] = end
        return await self._client.request(
            "GET", "/api/v1/billing/usage/breakdown", params=params if params else None
        )

    async def get_limits(self) -> dict[str, Any]:
        """Get current usage limits."""
        return await self._client.request("GET", "/api/v1/billing/limits")

    async def get_quotas(self) -> dict[str, Any]:
        """Get quota status for all resources."""
        return await self._client.request("GET", "/api/v1/quotas")

    async def export_usage(
        self,
        start: str | None = None,
        end: str | None = None,
        format: UsageExportFormat = "json",
    ) -> dict[str, Any] | str:
        """Export usage data as CSV or JSON."""
        params: dict[str, Any] = {"format": format}
        if start:
            params["start"] = start
        if end:
            params["end"] = end
        return await self._client.request("GET", "/api/v1/billing/usage/export", params=params)
